## Assignment Preparation

In [1]:
import pandas as pd

In [3]:
df = pd.read_csv('dataset.csv')

C:\Users\oguza\AppData\Local\Temp\ipykernel_53676\112859877.py:1: DtypeWarning: Columns (4) have mixed types. Specify dtype option on import or set low_memory=False.
  df = pd.read_csv('dataset.csv')


In [12]:
df.shape

(1083397, 42)

In [6]:
df.head()

restaurant_link       restaurant_name  \
0  g10001637-d10002227                Le 147   
1  g10001637-d14975787      Le Saint Jouvent   
2   g10002858-d4586832       Au Bout du Pont   
3   g10002986-d3510044   Le Relais de Naiade   
4   g10022428-d9767191  Relais Du MontSeigne   

                                   original_location country  \
0  ["Europe", "France", "Nouvelle-Aquitaine", "Ha...  France   
1  ["Europe", "France", "Nouvelle-Aquitaine", "Ha...  France   
2  ["Europe", "France", "Centre-Val de Loire", "B...  France   
3  ["Europe", "France", "Nouvelle-Aquitaine", "Co...  France   
4  ["Europe", "France", "Occitanie", "Aveyron", "...  France   

                region      province                      city  \
0   Nouvelle-Aquitaine  Haute-Vienne             Saint-Jouvent   
1   Nouvelle-Aquitaine  Haute-Vienne             Saint-Jouvent   
2  Centre-Val de Loire         Berry                Rivarennes   
3   Nouvelle-Aquitaine       Correze                   Lacelle   
4            Occitanie       Aveyron  Saint-Laurent-de-Levezou   

                                             address   latitude  longitude  \
0        10 Maison Neuve, 87510 Saint-Jouvent France  45.961674   1.169131   
1   16 Place de l Eglise, 87510 Saint-Jouvent France  45.957040   1.205480   
2           2 rue des Dames, 36800 Rivarennes France  46.635895   1.386133   
3  9 avenue Porte de la Correze 19170, 19170 Lace...  45.642610   1.824460   
4  route du Montseigne, 12620 Saint-Laurent-de-Le...  44.208860   2.960470   

   ... excellent very_good average poor terrible food service value  \
0  ...       2.0       0.0     0.0  0.0      0.0  4.0     4.5   4.0   
1  ...       2.0       2.0     1.0  0.0      0.0  NaN     NaN   NaN   
2  ...       3.0       1.0     0.0  0.0      0.0  NaN     NaN   NaN   
3  ...       1.0       0.0     0.0  0.0      0.0  4.5     4.5   4.5   
4  ...       4.0       7.0     0.0  0.0      0.0  4.5     4.5   4.5   

  atmosphere keywords  
0        NaN      NaN  
1        NaN      NaN  
2        NaN      NaN  
3        NaN      NaN  
4        NaN      NaN  

[5 rows x 42 columns]

In [5]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1083397 entries, 0 to 1083396
Data columns (total 42 columns):
 #   Column                             Non-Null Count    Dtype  
---  ------                             --------------    -----  
 0   restaurant_link                    1083397 non-null  object 
 1   restaurant_name                    1083397 non-null  object 
 2   original_location                  1083397 non-null  object 
 3   country                            1083397 non-null  object 
 4   region                             1033074 non-null  object 
 5   province                           742765 non-null   object 
 6   city                               682712 non-null   object 
 7   address                            1083397 non-null  object 
 8   latitude                           1067607 non-null  float64
 9   longitude                          1067607 non-null  float64
 10  claimed                            1081555 non-null  object 
 11  awards                  

## Data Cleaning
1. Identify the columns with mixed data types.
2. For each column, count the number of rows per data type.

In [110]:
def getDataTypes(dataframe):
    datatypes = {}

    for col in dataframe.columns:
        datatypes[col] = dataframe[col].apply(type).value_counts().to_dict()
        
    return pd.DataFrame(datatypes).fillna(0)


In [108]:
print(df.shape)
getDataTypes(df).T

(1083397, 42)


<class 'str'>  <class 'float'>
restaurant_link                        1083397.0              0.0
restaurant_name                        1083397.0              0.0
original_location                      1083397.0              0.0
country                                1083397.0              0.0
region                                 1033074.0          50323.0
province                                742765.0         340632.0
city                                    682712.0         400685.0
address                                1083397.0              0.0
latitude                                     0.0        1083397.0
longitude                                    0.0        1083397.0
claimed                                1081555.0           1842.0
awards                                  263133.0         820264.0
popularity_detailed                     988409.0          94988.0
popularity_generic                      985605.0          97792.0
top_tags                                972763.0         110634.0
price_level                             806192.0         277205.0
price_range                             304327.0         779070.0
meals                                   635347.0         448050.0
cuisines                                914294.0         169103.0
special_diets                           340256.0         743141.0
features                                317407.0         765990.0
vegetarian_friendly                    1083397.0              0.0
vegan_options                          1083397.0              0.0
gluten_free                            1083397.0              0.0
original_open_hours                     593832.0         489565.0
open_days_per_week                           0.0        1083397.0
open_hours_per_week                          0.0        1083397.0
working_shifts_per_week                      0.0        1083397.0
avg_rating                                   0.0        1083397.0
total_reviews_count                          0.0        1083397.0
default_language                        988204.0          95193.0
reviews_count_in_default_language            0.0        1083397.0
excellent                                    0.0        1083397.0
very_good                                    0.0        1083397.0
average                                      0.0        1083397.0
poor                                         0.0        1083397.0
terrible                                     0.0        1083397.0
food                                         0.0        1083397.0
service                                      0.0        1083397.0
value                                        0.0        1083397.0
atmosphere                                   0.0        1083397.0
keywords                                 99198.0         984199.0

3. Would removing missing values solve the mixed data type problem?

In [109]:
print(df.dropna().shape)
getDataTypes(df.dropna()).T

(2177, 42)


<class 'str'>  <class 'float'>
restaurant_link                           2177.0              0.0
restaurant_name                           2177.0              0.0
original_location                         2177.0              0.0
country                                   2177.0              0.0
region                                    2177.0              0.0
province                                  2177.0              0.0
city                                      2177.0              0.0
address                                   2177.0              0.0
latitude                                     0.0           2177.0
longitude                                    0.0           2177.0
claimed                                   2177.0              0.0
awards                                    2177.0              0.0
popularity_detailed                       2177.0              0.0
popularity_generic                        2177.0              0.0
top_tags                                  2177.0              0.0
price_level                               2177.0              0.0
price_range                               2177.0              0.0
meals                                     2177.0              0.0
cuisines                                  2177.0              0.0
special_diets                             2177.0              0.0
features                                  2177.0              0.0
vegetarian_friendly                       2177.0              0.0
vegan_options                             2177.0              0.0
gluten_free                               2177.0              0.0
original_open_hours                       2177.0              0.0
open_days_per_week                           0.0           2177.0
open_hours_per_week                          0.0           2177.0
working_shifts_per_week                      0.0           2177.0
avg_rating                                   0.0           2177.0
total_reviews_count                          0.0           2177.0
default_language                          2177.0              0.0
reviews_count_in_default_language            0.0           2177.0
excellent                                    0.0           2177.0
very_good                                    0.0           2177.0
average                                      0.0           2177.0
poor                                         0.0           2177.0
terrible                                     0.0           2177.0
food                                         0.0           2177.0
service                                      0.0           2177.0
value                                        0.0           2177.0
atmosphere                                   0.0           2177.0
keywords                                  2177.0              0.0

As it can be seen from the last table, removing missing values singularizes the datatypes for the object columns. But it makes no sense since the size of data shrinks from 1,083,397 to 2,177.

## Data Understanding
1. Are the review columns correlated with the rating columns? 

In [115]:
df[["excellent", "very_good", "average", "poor", "terrible", "total_reviews_count", "reviews_count_in_default_language"]]

excellent  very_good  average  poor  terrible  total_reviews_count  \
0              2.0        0.0      0.0   0.0       0.0                 36.0   
1              2.0        2.0      1.0   0.0       0.0                  5.0   
2              3.0        1.0      0.0   0.0       0.0                 13.0   
3              1.0        0.0      0.0   0.0       0.0                 34.0   
4              4.0        7.0      0.0   0.0       0.0                 11.0   
...            ...        ...      ...   ...       ...                  ...   
1083392        0.0        0.0      0.0   0.0       1.0                  2.0   
1083393        0.0        1.0      1.0   1.0       2.0                  6.0   
1083394        1.0        0.0      0.0   0.0       1.0                  2.0   
1083395        NaN        NaN      NaN   NaN       NaN                  0.0   
1083396        0.0        0.0      1.0   0.0       1.0                  3.0   

         reviews_count_in_default_language  
0                                      2.0  
1                                      5.0  
2                                      4.0  
3                                      1.0  
4                                     11.0  
...                                    ...  
1083392                                1.0  
1083393                                5.0  
1083394                                2.0  
1083395                                NaN  
1083396                                2.0  

[1083397 rows x 7 columns]

In [116]:
df[["food", "service", "value", "atmosphere", "avg_rating"]]

food  service  value  atmosphere  avg_rating
0         4.0      4.5    4.0         NaN         4.0
1         NaN      NaN    NaN         NaN         4.0
2         NaN      NaN    NaN         NaN         5.0
3         4.5      4.5    4.5         NaN         4.0
4         4.5      4.5    4.5         NaN         4.5
...       ...      ...    ...         ...         ...
1083392   NaN      NaN    NaN         NaN         2.5
1083393   NaN      NaN    NaN         NaN         2.0
1083394   NaN      NaN    NaN         NaN         3.0
1083395   NaN      NaN    NaN         NaN         NaN
1083396   NaN      NaN    NaN         NaN         1.5

[1083397 rows x 5 columns]

In [127]:
df[["excellent", "very_good", "average", "poor", "terrible", "total_reviews_count", "reviews_count_in_default_language", "food", "service", "value", "atmosphere", "avg_rating"]].corr().round(2)

excellent  very_good  average  poor  \
excellent                               1.00       0.84     0.73  0.63   
very_good                               0.84       1.00     0.93  0.83   
average                                 0.73       0.93     1.00  0.93   
poor                                    0.63       0.83     0.93  1.00   
terrible                                0.50       0.67     0.79  0.88   
total_reviews_count                     0.72       0.71     0.63  0.54   
reviews_count_in_default_language       0.96       0.95     0.89  0.81   
food                                    0.13       0.02    -0.06 -0.10   
service                                 0.13       0.02    -0.05 -0.09   
value                                   0.10      -0.01    -0.08 -0.13   
atmosphere                              0.17       0.07    -0.01 -0.06   
avg_rating                              0.09       0.01    -0.04 -0.08   

                                   terrible  total_reviews_count  \
excellent                              0.50                 0.72   
very_good                              0.67                 0.71   
average                                0.79                 0.63   
poor                                   0.88                 0.54   
terrible                               1.00                 0.45   
total_reviews_count                    0.45                 1.00   
reviews_count_in_default_language      0.67                 0.74   
food                                  -0.19                 0.06   
service                               -0.18                 0.04   
value                                 -0.20                -0.01   
atmosphere                            -0.14                 0.09   
avg_rating                            -0.16                 0.04   

                                   reviews_count_in_default_language  food  \
excellent                                                       0.96  0.13   
very_good                                                       0.95  0.02   
average                                                         0.89 -0.06   
poor                                                            0.81 -0.10   
terrible                                                        0.67 -0.19   
total_reviews_count                                             0.74  0.06   
reviews_count_in_default_language                               1.00  0.05   
food                                                            0.05  1.00   
service                                                         0.06  0.82   
value                                                           0.03  0.82   
atmosphere                                                      0.10  0.63   
avg_rating                                                      0.04  0.82   

                                   service  value  atmosphere  avg_rating  
excellent                             0.13   0.10        0.17        0.09  
very_good                             0.02  -0.01        0.07        0.01  
average                              -0.05  -0.08       -0.01       -0.04  
poor                                 -0.09  -0.13       -0.06       -0.08  
terrible                             -0.18  -0.20       -0.14       -0.16  
total_reviews_count                   0.04  -0.01        0.09        0.04  
reviews_count_in_default_language     0.06   0.03        0.10        0.04  
food                                  0.82   0.82        0.63        0.82  
service                               1.00   0.80        0.66        0.81  
value                                 0.80   1.00        0.56        0.77  
atmosphere                            0.66   0.56        1.00        0.63  
avg_rating                            0.81   0.77        0.63        1.00

When we have a look at the correlation matrix, we can see that the values for excellent reviewed listings have higher correlation ratio on rating columns. This means that if the number of excellent review increases for a listing, so does the rating columns. On the other hand, if we have a look at terrible reviews, we see that it has negative high correlation for the rating columns. This means that if the terrible review of a listing increases, the rating values decreases.The number of "excellent" review have higher impact that "very_good" reviews and so forth.

2. Are vegetarian-friendly restaurants better than non-vegetarian ones?

In [205]:
df['review_score'] = df[["excellent", "very_good", "average", "poor", "terrible", "total_reviews_count", "reviews_count_in_default_language"]].apply(lambda x: (x['excellent'] * 5 + x['very_good'] * 4 + x['average'] * 3 + x['poor'] * 2 + x['terrible'] * 1) / x['reviews_count_in_default_language'], axis = 1)

In [206]:
df[["excellent", "very_good", "average", "poor", "terrible", "total_reviews_count", "reviews_count_in_default_language", "review_score"]]

excellent  very_good  average  poor  terrible  total_reviews_count  \
0              2.0        0.0      0.0   0.0       0.0                 36.0   
1              2.0        2.0      1.0   0.0       0.0                  5.0   
2              3.0        1.0      0.0   0.0       0.0                 13.0   
3              1.0        0.0      0.0   0.0       0.0                 34.0   
4              4.0        7.0      0.0   0.0       0.0                 11.0   
...            ...        ...      ...   ...       ...                  ...   
1083392        0.0        0.0      0.0   0.0       1.0                  2.0   
1083393        0.0        1.0      1.0   1.0       2.0                  6.0   
1083394        1.0        0.0      0.0   0.0       1.0                  2.0   
1083395        NaN        NaN      NaN   NaN       NaN                  0.0   
1083396        0.0        0.0      1.0   0.0       1.0                  3.0   

         reviews_count_in_default_language  review_score  
0                                      2.0      5.000000  
1                                      5.0      4.200000  
2                                      4.0      4.750000  
3                                      1.0      5.000000  
4                                     11.0      4.363636  
...                                    ...           ...  
1083392                                1.0      1.000000  
1083393                                5.0      2.200000  
1083394                                2.0      3.000000  
1083395                                NaN           NaN  
1083396                                2.0      2.000000  

[1083397 rows x 8 columns]

In [207]:
df.groupby('vegetarian_friendly')["excellent", "very_good", "average", "poor", "terrible", "total_reviews_count", "reviews_count_in_default_language", "review_score", "food", "service", "value", "atmosphere", "avg_rating"].mean().T

C:\Users\oguza\AppData\Local\Temp\ipykernel_53676\2941516640.py:1: FutureWarning: Indexing with multiple keys (implicitly converted to a tuple of keys) will be deprecated, use a list instead.
  df.groupby('vegetarian_friendly')["excellent", "very_good", "average", "poor", "terrible", "total_reviews_count", "reviews_count_in_default_language", "review_score", "food", "service", "value", "atmosphere", "avg_rating"].mean().T


vegetarian_friendly                        N           Y
excellent                           8.475581   57.843392
very_good                           4.336071   23.116769
average                             1.817964    8.810135
poor                                1.063597    5.005333
terrible                            1.706817    5.515275
total_reviews_count                46.474440  226.332957
reviews_count_in_default_language  17.400029  100.290904
review_score                        4.025705    4.254182
food                                3.963376    4.235145
service                             3.935194    4.191907
value                               3.866922    4.091917
atmosphere                          3.781855    4.032328
avg_rating                          3.967643    4.176086

The definition of "better" is assumed to have higher reviews and ratings. If we have a look at the brief table above that showes the average review scores and ratings, we can see that vegetarian_friendly restaurants have higher review score and rating values.

3. Are there any significantly more expensive cuisines?

In [145]:
df.groupby('cuisines')['price_level'].value_counts()

cuisines                                         price_level
Afghani                                          €€-€€€         7
                                                 €              3
Afghani, Asian                                   €€-€€€         2
                                                 €              1
Afghani, Azerbaijani, Asian, Pakistani, Persian  €€-€€€         1
                                                               ..
Wine Bar, Tuscan, Central-Italian                €€-€€€         1
                                                 €€€€           1
Wine Bar, Tuscan, Central-Italian, Deli          €€-€€€         1
Wine Bar, Tuscan, Romana, Central-Italian        €€-€€€         1
Yunnan                                           €€-€€€         1
Name: price_level, Length: 103277, dtype: int64

In [227]:
df['cuisines']

0                                                     French
1                                                        NaN
2                                           French, European
3                                                     French
4                                                     French
                                 ...                        
1083392                                                  NaN
1083393                                     Eastern European
1083394                                                  NaN
1083395    French, American, Bar, International, European...
1083396                                                Diner
Name: cuisines, Length: 1083397, dtype: object

In [153]:
df['cuisines'].str.split(',', expand=True).stack()

0        0            French
2        0            French
         1          European
3        0            French
4        0            French
                   ...      
1083395  3     International
         4          European
         5               Pub
         6          Romanian
1083396  0             Diner
Length: 2232933, dtype: object

In [158]:
stacked_cuisines = df.set_index('price_level').cuisines.str.split(',', expand=True).stack().reset_index('price_level')
stacked_cuisines.columns = ['price_level', 'cuisines']

In [224]:
stacked_cuisines

price_level        cuisines
0            €          French
0            €          French
1            €        European
0            €          French
0       €€-€€€          French
..         ...             ...
3            €   International
4            €        European
5            €             Pub
6            €        Romanian
0          NaN           Diner

[2232933 rows x 2 columns]

In [229]:
counts = stacked_cuisines.groupby('cuisines').count()
counts.columns = ['total_counts']
counts = counts.reset_index()

In [245]:
counts.set_index('cuisines')

total_counts
cuisines                
 Afghani             402
 African             583
 Albanian            232
 Algerian            155
 American           8697
...                  ...
Venezuelan           138
Vietnamese          1158
Welsh                 61
Wine Bar            1935
Yunnan                 1

[306 rows x 1 columns]

In [181]:
tmp = pd.DataFrame(stacked_cuisines.groupby('cuisines')['price_level'].value_counts())
tmp.columns = ['counts']
tmp = tmp.reset_index()

In [248]:
tmp = tmp[tmp['price_level'] == '€€€€'].set_index('cuisines').join(counts.set_index('cuisines'))

In [250]:
tmp['avg_rate'] = tmp['counts'] / tmp['total_counts']

In [253]:
tmp.sort_values('counts', ascending = False)

price_level  counts  total_counts  avg_rate
cuisines                                                           
 European                      €€€€   10178        157158  0.064763
French                         €€€€    6275         77743  0.080715
 Mediterranean                 €€€€    6228        121612  0.051212
Italian                        €€€€    6166        180766  0.034110
 Seafood                       €€€€    4266         56202  0.075905
...                             ...     ...           ...       ...
Japanese sweets parlour        €€€€       1             6  0.166667
Medicinal foods                €€€€       1            10  0.100000
 Salvadoran                    €€€€       1            16  0.062500
Persian                        €€€€       1           118  0.008475
 Singaporean                   €€€€       1            95  0.010526

[249 rows x 4 columns]

In [252]:
tmp.sort_values('avg_rate', ascending = False)

price_level  counts  total_counts  avg_rate
cuisines                                                   
 Shojin                €€€€       1             2  0.500000
 Kappo                 €€€€       2             4  0.500000
Beijing cuisine        €€€€       1             3  0.333333
 Kyoto cuisine         €€€€       2            11  0.181818
Japanese Fusion        €€€€       8            46  0.173913
...                     ...     ...           ...       ...
 Nepalese              €€€€       3           868  0.003456
Soups                  €€€€       1           298  0.003356
Fast food              €€€€      60         19596  0.003062
Street Food            €€€€       3          1013  0.002962
 Balti                 €€€€      13          5598  0.002322

[249 rows x 4 columns]

The value of counts for European cuisine is almost doubles the following higest which is French. However if we look at the average rate that is number of highest price rank over overall listing counts, we cannot say certain cuisines are more expensive than the others. For example for Shojin cuisine 1 of 2 listings is ranked €€€€ which is 50%. But with the amount of little listings, it cannot be said that this is a costy cuisine.

## Business-specific

1. In the assets directory, you will see a very small dataset called europe_capitals_population_and_area.csv. A gluten-free restaurant wants to open a new restaurant in a European capital where gluten-free restaurants are underrepresented. Assuming there are no other factors, except population and gluten-free restaurant density, what would be the top 5 capitals to open that restaurant?

In [261]:
df.head()

restaurant_link       restaurant_name  \
0  g10001637-d10002227                Le 147   
1  g10001637-d14975787      Le Saint Jouvent   
2   g10002858-d4586832       Au Bout du Pont   
3   g10002986-d3510044   Le Relais de Naiade   
4   g10022428-d9767191  Relais Du MontSeigne   

                                   original_location country  \
0  ["Europe", "France", "Nouvelle-Aquitaine", "Ha...  France   
1  ["Europe", "France", "Nouvelle-Aquitaine", "Ha...  France   
2  ["Europe", "France", "Centre-Val de Loire", "B...  France   
3  ["Europe", "France", "Nouvelle-Aquitaine", "Co...  France   
4  ["Europe", "France", "Occitanie", "Aveyron", "...  France   

                region      province                      city  \
0   Nouvelle-Aquitaine  Haute-Vienne             Saint-Jouvent   
1   Nouvelle-Aquitaine  Haute-Vienne             Saint-Jouvent   
2  Centre-Val de Loire         Berry                Rivarennes   
3   Nouvelle-Aquitaine       Correze                   Lacelle   
4            Occitanie       Aveyron  Saint-Laurent-de-Levezou   

                                             address   latitude  longitude  \
0        10 Maison Neuve, 87510 Saint-Jouvent France  45.961674   1.169131   
1   16 Place de l Eglise, 87510 Saint-Jouvent France  45.957040   1.205480   
2           2 rue des Dames, 36800 Rivarennes France  46.635895   1.386133   
3  9 avenue Porte de la Correze 19170, 19170 Lace...  45.642610   1.824460   
4  route du Montseigne, 12620 Saint-Laurent-de-Le...  44.208860   2.960470   

   ... very_good average poor terrible food service value atmosphere keywords  \
0  ...       0.0     0.0  0.0      0.0  4.0     4.5   4.0        NaN      NaN   
1  ...       2.0     1.0  0.0      0.0  NaN     NaN   NaN        NaN      NaN   
2  ...       1.0     0.0  0.0      0.0  NaN     NaN   NaN        NaN      NaN   
3  ...       0.0     0.0  0.0      0.0  4.5     4.5   4.5        NaN      NaN   
4  ...       7.0     0.0  0.0      0.0  4.5     4.5   4.5        NaN      NaN   

  review_score  
0     5.000000  
1     4.200000  
2     4.750000  
3     5.000000  
4     4.363636  

[5 rows x 43 columns]

In [278]:
pd.DataFrame(df[df['gluten_free'] == 'Y'].groupby(['country', 'city'])['restaurant_link'].count()).reset_index()

country           city  restaurant_link
0      Austria        Abtenau                1
1      Austria     Achenkirch                2
2      Austria           Aich                2
3      Austria  Aigen-Schlagl                1
4      Austria        Alpbach                2
...        ...            ...              ...
11628    Wales  Wolf's Castle                2
11629    Wales        Wrexham               24
11630    Wales    Y Felinheli                3
11631    Wales  Ystrad Mynach                2
11632    Wales  Ystradgynlais                3

[11633 rows x 3 columns]

In [349]:
data = pd.read_csv('europe_capitals_population_and_area.csv', sep = ';')

In [284]:
alldata = pd.merge(data, pd.DataFrame(df[df['gluten_free'] == 'Y'].groupby(['country', 'city'])['restaurant_link'].count()).reset_index(), left_on=['Country', 'City'], right_on=['country', 'city'], how='left').fillna(0)

In [329]:
alldata['restaurant_link_score'] = pd.cut(alldata.restaurant_link, bins = 10, labels = [10, 9, 8, 7, 6, 5, 4, 3, 2, 1]).astype(int)

In [328]:
alldata['area_score'] = pd.cut(alldata['Population']/alldata['Area in km2 '].str.replace(',', '.').astype(float), bins = 10, labels = [1, 2, 3, 4, 5, 6, 7, 8, 9, 10]).astype(int)

In [344]:
alldata['overall_score'] = (alldata['restaurant_link_score'] + alldata['area_score']) / 2

In [345]:
alldata

CityNo        City      Country  Population Area in km2    country  \
0        1   Amsterdam   Netherland      851573       219,32         0   
1        2      Athens       Greece      664046         2929    Greece   
2        4   Barcelona        Spain     1620809        101,4         0   
3        4      Berlin      Germany     3711930        891,7         0   
4        5  Bratislava     Slovakia      424428        367,6  Slovakia   
5        6    Brussels      Belgium     1191604       161,38         0   
6        7    Budapest      Hungary     1779361        525,2   Hungary   
7        8  Copenhagen      Denmark      602481        88,25   Denmark   
8        9      Dublin      Ireland      527612          115   Ireland   
9       10   Edinburgh     Scotland      482005          264         0   
10      11      Geneva  Switzerland      198979        15,92         0   
11      12     Hamburg      Germany     1822445          755         0   
12      13    Helsinki      Finland      631695        184,5   Finland   
13      14      Krakow       Poland      765320          327         0   
14      15      Lisbon     Portugal      504718          100  Portugal   
15      16   Ljubljana     Slovenia      279631        163,8         0   
16      17      London      England     8825000          607         0   
17      18  Luxembourg   Luxembourg      590667       2586,4         0   
18      19        Lyon       France      506615        47,87    France   
19      20      Madrid        Spain     3141991        604,3     Spain   
20      21       Milan        Italy     1370074       181,76     Italy   
21      22      Munich      Germany     1456039       310,43   Germany   
22      23      Oporto     Portugal      214349        41,42         0   
23      24        Oslo       Norway      634293          454         0   
24      25       Paris       France     2206488        105,4    France   
25      26      Prague        Czech     1301132          298         0   
26      27        Rome        Italy     2868782         1285     Italy   
27      28   Stockholm       Sweden      952058          188    Sweden   
28      29      Vienna      Austria     1889083        414,6   Austria   
29      30      Warsaw       Poland     1764615          517         0   
30      31      Zurich  Switzerland      409241        87,93         0   

          city  restaurant_link  restaurant_link_score  area_score  \
0            0              0.0                     10           2   
1       Athens            271.0                      9           1   
2            0              0.0                     10           8   
3            0              0.0                     10           2   
4   Bratislava             67.0                     10           1   
5            0              0.0                     10           4   
6     Budapest            433.0                      8           2   
7   Copenhagen            251.0                      9           4   
8       Dublin            596.0                      7           3   
9            0              0.0                     10           1   
10           0              0.0                     10           6   
11           0              0.0                     10           2   
12    Helsinki            199.0                      9           2   
13           0              0.0                     10           2   
14      Lisbon            537.0                      7           3   
15           0              0.0                     10           1   
16           0              0.0                     10           7   
17           0              0.0                     10           1   
18        Lyon             54.0                     10           6   
19      Madrid           1368.0                      2           3   
20       Milan           1154.0                      4           4   
21      Munich            256.0                      9           3   
22           0           

In [347]:
alldata.sort_values('overall_score', ascending=False).head(5)

CityNo       City      Country  Population Area in km2  country   city  \
2        4  Barcelona        Spain     1620809        101,4       0      0   
24      25      Paris       France     2206488        105,4  France  Paris   
16      17     London      England     8825000          607       0      0   
10      11     Geneva  Switzerland      198979        15,92       0      0   
18      19       Lyon       France      506615        47,87  France   Lyon   

    restaurant_link  restaurant_link_score  area_score  overall_score  
2               0.0                     10           8            9.0  
24            582.0                      7          10            8.5  
16              0.0                     10           7            8.5  
10              0.0                     10           6            8.0  
18             54.0                     10           6            8.0

My top 5 capitals would be Barcelona, Paris, London, Geneva and Lyon. I cut the number of restaurant_link and population for each km2 into 10 groups. I ranked them and calculated the overall score for each capital. Sorting them in descending order gave the top capitals.

2. Think and propose a couple of other ways this dataset could be used to help businesses.

In [350]:
data

CityNo        City      Country  Population Area in km2 
0        1   Amsterdam   Netherland      851573       219,32
1        2      Athens       Greece      664046         2929
2        4   Barcelona        Spain     1620809        101,4
3        4      Berlin      Germany     3711930        891,7
4        5  Bratislava     Slovakia      424428        367,6
5        6    Brussels      Belgium     1191604       161,38
6        7    Budapest      Hungary     1779361        525,2
7        8  Copenhagen      Denmark      602481        88,25
8        9      Dublin      Ireland      527612          115
9       10   Edinburgh     Scotland      482005          264
10      11      Geneva  Switzerland      198979        15,92
11      12     Hamburg      Germany     1822445          755
12      13    Helsinki      Finland      631695        184,5
13      14      Krakow       Poland      765320          327
14      15      Lisbon     Portugal      504718          100
15      16   Ljubljana     Slovenia      279631        163,8
16      17      London      England     8825000          607
17      18  Luxembourg   Luxembourg      590667       2586,4
18      19        Lyon       France      506615        47,87
19      20      Madrid        Spain     3141991        604,3
20      21       Milan        Italy     1370074       181,76
21      22      Munich      Germany     1456039       310,43
22      23      Oporto     Portugal      214349        41,42
23      24        Oslo       Norway      634293          454
24      25       Paris       France     2206488        105,4
25      26      Prague        Czech     1301132          298
26      27        Rome        Italy     2868782         1285
27      28   Stockholm       Sweden      952058          188
28      29      Vienna      Austria     1889083        414,6
29      30      Warsaw       Poland     1764615          517
30      31      Zurich  Switzerland      409241        87,93

* With the help of capitals population data, many marketing strategies can be conducted. For example for the restaurants that has a "Reservations" feature and has a higer that 4.0 rating, close to peak hours, a notification can be sent to users in order them to make reservations.
* Same analysis with gluten-free option can be made to vegetarian and vegan options. For the places with underrepresented vegetarian_friendly places, new businesses can be promoted.
* By the assumption of the capital cities have higher number of workers, the places with the original open hours including lunch brakes can be promoted for those users for a lunch brake maybe with a bit of a coupon if applicable.